docs/tools/builder/index.rst

```diff
@@ -0,0 +1,329 @@
+    +--------------------+------------------------------------------------------------+
+
+
+.. autofunction:: eth.tools.builder.state.expect
+
+    This specifies the expected result of the test.
+
+    For state tests, multiple expectations can be given, differing in the the transaction data, gas
+    limit, and value, in the applicable networks, and as a result also in the post state. VM tests
+    support only a single expectation with no specified network and no transaction (here, its role is
+    played by :func:`~eth.tools.builder.state.execution`).
+
+    * ``post_state`` is a list of state definition in the same form as expected by `pre_state`. State items
+      that are not set explicitly default to their pre state.
+
+    * ``networks`` defines the forks under which the expectation is applicable. It should be a sublist of
+      the following identifiers (also available in `ALL_FORKS`):
+
+      * ``"Frontier"``
+      * ``"Homestead"``
+      * ``"EIP150"``
+      * ``"EIP158"``
+      * ``"Byzantium"``
+
+    * ``transaction`` is a dictionary coming in two variants. For the main shard:
+
+      +----------------+-------------------------------+
+      | key            | description                   |
+      +================+===============================+
+      | ``"data"``     | the transaction data,         |
+      +----------------+-------------------------------+
+      | ``"gasLimit"`` | the transaction gas limit,    |
+      +----------------+-------------------------------+
+      | ``"gasPrice"`` | the gas price,                |
+      +----------------+-------------------------------+
+      | ``"nonce"``    | the transaction nonce,        |
+      +----------------+-------------------------------+
+      | ``"value"``    | the transaction value         |
+      +----------------+-------------------------------+
+
+    In addition, one should specify either the signature itself (via keys ``"v"``, ``"r"``, and ``"s"``) or
+    a private key used for signing (via ``"secretKey"``).
+
+
+Chain Builder
+-------------
+
+The chain builder utils are intended to reduce common boilerplace for both
+construction of chain classes as well as building up some desired chain state.
+
+These tools are best used in conjunction with ``cytoolz.pipe``.
+
+
+Constructing Chain Classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following utilities are provided to assist with constructing a chain class.
+
+
+.. autofunction:: eth.tools.builder.chain.fork_at
+
+    Adds the ``vm_class`` to the chain's ``vm_configuration``.  The follow
+
+    .. code-block:: python
+
+        from cytoolz import pipe
+        from eth.chains.base import MiningChain
+        from eth.tools.builder.chain import fork_at
+
+        FrontierOnlyChain = pipe(MiningChain, fork_at(FrontierVM, 0))
+
+        # these two classes are functionally equivalent.
+        class FrontierOnlyChain(MiningChain):
+            vm_configuration = (
+                (0, FrontierVM),
```

Comment:
Might be nice to flip the args in `fork_at` just for consistency with this format.

I assume it is in the current order for currying, but we could do a kwarg curry.